### PR TITLE
RCHAIN-2729: Return an error if phloLimit passed to deploy command is less than intrinsic phlo

### PR DESCRIPTION
**Pull request by JosDenmark**
## Overview
This PR changes `RuntimeInterpreter` to throw an error if the deploy being evaluated provides phlo less than the minimum cost of evaluating a deploy. It is the first of a series of validity checks that deploys will be subject to.

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-2729

### CODE REVIEW

**Comment by JosDenmark** [comment]
bors try

**Comment by JosDenmark** [comment]
> We want to replay deploys with an insufficient phlo - otherwise a malicious validator can mark a bunch of valid transactions as invalid.

I'm unsure what you mean. The insufficient phlo check is done in `doInj`, which is used both in `doReplayEval` and `doEval`. The phlo check is repeated independently during replay.

**Comment by KentShikama** [comment]
`   "replayComputeState" should "not replay a deploy that throws an InsufficientPhloError" in {`

I interpreted that test title to mean that we don't replay deploys with an insufficient phlo.

**Comment by JosDenmark** [comment]
Ah. I see. I'll rename to `"replayComputeState" should "rethrow InsufficientPhloError"`.

**Comment by JosDenmark** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala [hunk 1]
```diff
@@ -293,19 +291,27 @@ class RuntimeManagerImpl[F[_]: Concurrent] private[rholang] (
...
     implicit val rand: Blake2b512Random = Blake2b512Random(
       DeployData.toByteArray(ProtoUtil.stripDeployData(deploy))
     )
-    Interpreter[F].injAttempt(
-      reducer,
-      errorLog,
-      deploy.term,
-      Cost(deploy.phloLimit)
-    )
+    val intrinsicGas: Long = parsingCost(deploy.term).value
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: Why not just `parsingCost` or `parsingGas`?

File: casper/src/main/scala/coop/rchain/casper/util/SignatureAlgorithms.scala [hunk 2]
```diff
@@ -4,10 +4,19 @@ import coop.rchain.crypto.signatures.{Ed25519, Secp256k1}
...
+  def lookupSigAlg(algorithm: String): SignatureAlgorithm =
     algorithm match {
       case "ed25519"   => Ed25519.sign
       case "secp256k1" => Secp256k1.sign
       case _           => throw new Exception(s"Unknown signature algorithm $algorithm")
     }
+
+  type SignatureVerifyAlgorithm = (Array[Byte], Array[Byte], Array[Byte]) => Boolean
+  @SuppressWarnings(Array("org.wartremover.warts.Throw")) // TODO remove throw
+  def lookupSigVerifyAlg(algorithm: String): SignatureVerifyAlgorithm =
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/SignatureAlgorithms.scala]: So where do you use this?

File: casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala [hunk 3]
```diff
@@ -508,6 +508,13 @@ object ProtoUtil {
       phloLimit = phlos
     )
 
+  def sourceDeployNowWithPhlo(source: String, phlos: Long): DeployData =
+    sourceDeploy(
+      source,
+      System.currentTimeMillis(),
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: I prefer if we could use `Time[F].currentMillis` so we can inject logicalTime for tests

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala (hunk 1 above)
Comment (JosDenmark) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: In the future, it's possible that it includes costs besides parsing. For example, while we don't have charging in the normalizer, we might include a cost for normalization.

File: casper/src/main/scala/coop/rchain/casper/util/SignatureAlgorithms.scala (hunk 2 above)
Comment (JosDenmark) [reply casper/src/main/scala/coop/rchain/casper/util/SignatureAlgorithms.scala]: This was part of a larger PR at first which used it. It should be removed.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala (hunk 1 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: I see; I say we rename it then

Comment (JosDenmark) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: Rename it to what? The terminology is taken from Ethereum, which has a somewhat analogous minimum.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: To intrinsic gas

**Pull request closed without merging**